### PR TITLE
Implement DELETE /v1/chains/:chainId/notifications/devices/:uuid/safes/:safeAddress

### LINKED ISSUE #260: Add DELETE /v1/chains/<chain_id>/safes/<safe_address>/delegates/<delegate_address>

### LINKED ISSUE #261: Add DELETE /v1/chains/<chain_id>/notifications/devices/<uuid>/safes/<safe_address>

### PR DESCRIPTION
**Pull request by hectorgomezv**
(Please review #267 first)

Closes #261 

Add the DELETE endpoint `/v1/chains/:chainId/notifications/devices/:uuid/safes/:safeAddress`
- OpenAPI specification was added.
- Unit (controller) tests were added.

### CODE REVIEW

File: src/datasources/transaction-api/transaction-api.service.ts [hunk 1]
```diff
@@ -436,4 +436,13 @@ export class TransactionApi implements ITransactionApi {
       throw this.httpErrorFactory.from(error);
     }
   }
+
+  async deleteDeviceRegistration(uuid: string, safeAddress: string) {
```
Comment (fmrsabino) [inline src/datasources/transaction-api/transaction-api.service.ts:440]: This is returning a `NetworkResponse` to the domain layer right?

I'd not expose this to the domain/route layers.

File: src/domain/notifications/notifications.repository.interface.ts [hunk 2]
```diff
@@ -14,4 +14,16 @@ export interface INotificationsRepository {
...
     safeRegistration: SafeRegistration,
   ): Promise<SafeRegistration>;
+
+  /**
+   * Deletes a a notification target, identified by its client {@link uuid}.
+   * @param chainId
+   * @param uuid
+   * @param safeAddress
+   */
+  deleteDevice(
```
Comment (fmrsabino) [inline src/domain/notifications/notifications.repository.interface.ts:24]: nitpick: Maybe `unregisterDevice` is more meaningful. We have `deleteDeviceRegistration` on the API layer which I think is a good naming choice.

File: src/routes/notifications/notifications.controller.spec.ts [hunk 3]
```diff
@@ -237,4 +241,78 @@ describe('Notifications Controller (Unit)', () => {
...
+    it('Success', async () => {
+      const uuid = faker.datatype.uuid();
+      const safeAddress = faker.finance.ethereumAddress();
+      mockNetworkService.get.mockImplementation((url) =>
+        url.includes(`${safeConfigUrl}/api/v1/chains/`)
+          ? Promise.resolve({ data: chainBuilder().build() })
+          : rejectForUrl(url),
+      );
+      mockNetworkService.delete.mockImplementation((url) =>
+        url.includes(
```
Comment (fmrsabino) [inline src/routes/notifications/notifications.controller.spec.ts:255]: Why do we use `includes` here? We should know which URL to match or?

File: src/routes/notifications/notifications.controller.spec.ts [hunk 4]
```diff
@@ -237,4 +241,78 @@ describe('Notifications Controller (Unit)', () => {
...
+        )
+          ? Promise.resolve()
+          : rejectForUrl(url),
+      );
+
+      await request(app.getHttpServer())
+        .delete(
+          `/chains/${faker.random.numeric()}/notifications/devices/${uuid}/safes/${safeAddress}`,
+        )
+        .expect(200);
```
Comment (fmrsabino) [inline src/routes/notifications/notifications.controller.spec.ts:266]: What is the expected `body` here?

File: src/routes/notifications/notifications.controller.spec.ts [hunk 5]
```diff
@@ -237,4 +241,78 @@ describe('Notifications Controller (Unit)', () => {
...
+      expect(mockNetworkService.delete).toBeCalledTimes(1);
```
Comment (fmrsabino) [inline src/routes/notifications/notifications.controller.spec.ts:267]: I'd add the payload expected to be used for this call.

File: src/routes/notifications/notifications.controller.ts [hunk 6]
```diff
@@ -16,4 +23,14 @@ export class NotificationsController {
   ): Promise<void> {
     return this.notificationsService.registerDevice(registerDeviceDto);
   }
+
+  @ApiOkResponse()
```
Comment (fmrsabino) [inline src/routes/notifications/notifications.controller.ts:27]: Why do we need to decorate the route with `@ApiOkResponse()`?

File: src/routes/notifications/notifications.service.ts [hunk 7]
```diff
@@ -76,6 +76,27 @@ export class NotificationsService {
     }
   }
 
+  /**
+   * Sends a delete request for a chain, notification uuid and safe address.
```
Comment (fmrsabino) [inline src/routes/notifications/notifications.service.ts:80]: Not sure if I completely get this sentence 😅 Is it enumerating the parameters?

File: src/routes/notifications/notifications.service.ts [hunk 8]
```diff
@@ -76,6 +76,27 @@ export class NotificationsService {
...
+   * The uuid is expected to be managed by the client. Its value should be equal
+   * to the provided when the client called {@link registerDevice}.
```
Comment (fmrsabino) [inline src/routes/notifications/notifications.service.ts:82]:
```suggestion
   * The uuid is expected to be managed by the client. Its value should be equal
   * to the one provided when the client called {@link registerDevice}.
```

File: src/routes/notifications/notifications.service.ts [hunk 9]
```diff
@@ -76,6 +76,27 @@ export class NotificationsService {
...
+   *
+   * @param chainId
+   * @param uuid the same uuid provided when the device was registered.
+   * @param safeAddress
```
Comment (fmrsabino) [inline src/routes/notifications/notifications.service.ts:86]: I think you need to add a description for each parameter for it to show up in the documentation. It doesn't mention anything here though: https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns

Tooltip example (from IntelliJ)
image

File: src/routes/notifications/notifications.controller.ts (hunk 6 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.controller.ts:27]: No reason, in this case Nest will default to status 200 👍🏻

File: src/routes/notifications/notifications.service.ts (hunk 7 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.service.ts:80]: Yes, that was the intention but it's poorly written and redundant also 🙂 I will remove it.

File: src/domain/notifications/notifications.repository.interface.ts (hunk 2 above)
Comment (hectorgomezv) [reply src/domain/notifications/notifications.repository.interface.ts]: Changed in [d880fcb](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/d880fcb3732f5bd60894507bcb3104e4a46b7714)

File: src/routes/notifications/notifications.controller.spec.ts (hunk 4 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.controller.spec.ts]: An empty body, changed the expectation in [d880fcb](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/d880fcb3732f5bd60894507bcb3104e4a46b7714)

File: src/routes/notifications/notifications.controller.spec.ts (hunk 5 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.controller.spec.ts:266]: Done [d880fcb](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/d880fcb3732f5bd60894507bcb3104e4a46b7714)

File: src/routes/notifications/notifications.controller.ts (hunk 6 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.controller.ts]: Changed in [d880fcb](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/d880fcb3732f5bd60894507bcb3104e4a46b7714)

File: src/routes/notifications/notifications.service.ts (hunk 7 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.service.ts]: Fixed in [d880fcb](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/d880fcb3732f5bd60894507bcb3104e4a46b7714)

File: src/routes/notifications/notifications.service.ts (hunk 8 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.service.ts]: Thanks! Changed in [d880fcb](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/d880fcb3732f5bd60894507bcb3104e4a46b7714)

File: src/routes/notifications/notifications.service.ts (hunk 9 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.service.ts]: Visual Studio Code shows the complete tooltip, but I changed it anyway in [d880fcb](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/d880fcb3732f5bd60894507bcb3104e4a46b7714) 👍 

Screenshot 2023-02-10 at 14 44 41

File: src/routes/notifications/notifications.service.ts (hunk 7 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.service.ts]: Rephrased in [77a2868](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/77a2868a5ebb4ce4774cb16efc978618b404e525)

File: src/routes/notifications/notifications.controller.spec.ts [hunk 10]
```diff
@@ -237,4 +241,71 @@ describe('Notifications Controller (Unit)', () => {
...
   });
+
+  describe('DELETE /chains/:chainId/notifications/devices/:uuid/safes/:safeAddress', () => {
+    it('Success', async () => {
+      const uuid = faker.datatype.uuid();
+      const safeAddress = faker.finance.ethereumAddress();
+      const chain = chainBuilder().build();
+      const expectedProviderURL = `${chain.transactionService}/api/v1/notifications/devices/${uuid}/safes/${safeAddress}`;
+      mockNetworkService.get.mockImplementation((url) =>
+        url.includes(`${safeConfigUrl}/api/v1/chains/`)
```
Comment (fmrsabino) [inline src/routes/notifications/notifications.controller.spec.ts:252]: Same as https://github.com/5afe/safe-client-gateway-nest/pull/268#discussion_r1102693845

File: src/routes/notifications/notifications.controller.spec.ts [hunk 11]
```diff
@@ -237,4 +241,71 @@ describe('Notifications Controller (Unit)', () => {
...
+        .expect({});
+      expect(mockNetworkService.delete).toBeCalledTimes(1);
+      expect(mockNetworkService.delete).toBeCalledWith(expectedProviderURL);
+    });
+
+    it('Failure: Config API fails', async () => {
+      const uuid = faker.datatype.uuid();
+      const safeAddress = faker.finance.ethereumAddress();
+      mockNetworkService.get.mockImplementation((url) =>
+        url.includes(`${safeConfigUrl}/api/v1/chains/`)
```
Comment (fmrsabino) [inline src/routes/notifications/notifications.controller.spec.ts:274]: Same as https://github.com/5afe/safe-client-gateway-nest/pull/268#discussion_r1102693845

File: src/routes/notifications/notifications.controller.spec.ts [hunk 12]
```diff
@@ -237,4 +241,71 @@ describe('Notifications Controller (Unit)', () => {
...
+        .expect(503);
+      expect(mockNetworkService.delete).toBeCalledTimes(0);
+    });
+
+    it('Failure: Transaction API fails', async () => {
+      const uuid = faker.datatype.uuid();
+      const safeAddress = faker.finance.ethereumAddress();
+      const chain = chainBuilder().build();
+      mockNetworkService.get.mockImplementation((url) =>
+        url.includes(`${safeConfigUrl}/api/v1/chains/`)
```
Comment (fmrsabino) [inline src/routes/notifications/notifications.controller.spec.ts:292]: Same as https://github.com/5afe/safe-client-gateway-nest/pull/268#discussion_r1102693845

File: src/routes/notifications/notifications.service.ts (hunk 9 above)
Comment (fmrsabino) [reply src/routes/notifications/notifications.service.ts]: Interesting. Yeah now I see the params correctly.

image

No big issue though! Thanks for updating!

**Review by fmrsabino — COMMENTED** [review]
Looks good 👏 

Just some additional `includes` that are present in some tests.

File: src/routes/notifications/notifications.controller.spec.ts (hunk 10 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.controller.spec.ts]: Changed in [20a87a3](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/20a87a357afa6c9e9059a2e259a0a64d3468d812)

File: src/routes/notifications/notifications.controller.spec.ts (hunk 11 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.controller.spec.ts]: Changed in [20a87a3](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/20a87a357afa6c9e9059a2e259a0a64d3468d812)

File: src/routes/notifications/notifications.controller.spec.ts (hunk 12 above)
Comment (hectorgomezv) [reply src/routes/notifications/notifications.controller.spec.ts]: Changed in [20a87a3](https://github.com/5afe/safe-client-gateway-nest/pull/268/commits/20a87a357afa6c9e9059a2e259a0a64d3468d812)

**Review by fmrsabino — APPROVED** [review]

**Pull request merged**